oauth2client/contrib/gce.py

```diff
@@ -85,32 +121,110 @@ class AppAssertionCredentials(AssertionCredentials):
+                the actual scopes for tokens requested. These scopes are
+                set at VM instance creation time and won't change.
+            service_account_email:
+                the email for these credentials. This can be used with custom
+                service accounts, or left blank to use the default service
+                account for the instance. Usually the compute engine service
+                account.
+            service_account_info:
+                Deserialized JSON object, returned by self.service_account_info
         """
+
+        self._service_account_info = service_account_info or {
+            'email': service_account_email
+        }
+        self._project_id = None
+        self._partial = service_account_info is None
+
         if scope:
-            warnings.warn(_SCOPES_WARNING)
-        # This is just provided for backwards compatibility, but is not
-        # used by this class.
-        self.scope = util.scopes_to_string(scope)
+            if self.has_scopes(scope):
+                warnings.warn(_SCOPES_WARNING)
+            else:
+                raise ValueError(_SCOPES_WARNING)
+
         self.kwargs = kwargs
 
         # Assertion type is no longer used, but still in the
         # parent class signature.
         super(AppAssertionCredentials, self).__init__(None)
-        self._service_account_email = None
 
-    @classmethod
-    def from_json(cls, json_data):
-        data = json.loads(_from_bytes(json_data))
-        return AppAssertionCredentials(data['scope'])
+    @property
+    def service_account_info(self):
+        """Info about this service account
+        By using _get_service_account_info this property is
+        always guaranteed to have the members ['email', 'scopes'].
+        It may also have the member: 'aliases'
+        """
+        return self._get_service_account_info()
+
+    @property
+    def scopes(self):
+        return self.service_account_info['scopes']
+
+    @scopes.setter
+    def scopes(self, value):
+        pass
+
+    @property
+    def service_account_email(self):
+        return self.service_account_info['email']
+
+    @property
+    def project_id(self):
+        if not self._project_id:
+            self._project_id = _get_metadata(
+                path=['project', 'project-id'],
+                recursive=False
+            )
+        return self._project_id
+
+    @property
+    def serialization_data(self):
+        return {'service_account_info': self.service_account_info}
+
+    def _get_service_account_info(self, http_request=None):
+        """Retrieves the full info for a service account and caches it.
+        Args:
+            http_request: an httplib2.Http().request object or equivalent
+                with which to make the call to the metadata server
+        Returns:
+            A deserialized JSON service account object of the form:
+                {
+                   'aliases': [],
+                   'scopes': [],
```

Comment:
This implies that `scopes` and `aliases` will be empty lists. Is that always the case?